keps/sig-auth/541-external-credential-providers/README.md

```diff
@@ -313,20 +337,57 @@ has completed parsing its `kubeconfig`, flags and environment variables).
+`"k8s.io/client-go/tools/clientcmd/api/v1".Cluster`), but also allows arbitrary
 per-cluster configuration to be passed to the executable via the `config` field.
 This field can contain arbitrary data that is passed to the executable without
 modification.  This allows extra user-defined data (i.e. an OAuth client ID for
 audience scoping) to be passed through the `spec.cluster` field.  The user
-configures this via the `kubeconfig`'s `clusters.[...].cluster.extensions[exec].extension`
-field.  The `exec` named extension is reserved for this purpose.
+configures this via the `kubeconfig`'s `clusters.[...].cluster.extensions[client.authentication.k8s.io/exec].extension`
+field.  The `client.authentication.k8s.io/exec` named extension is reserved for this
+purpose.
+
+To ensure that the `Cluster` struct maintains the same cluster connection
+capabilities as one gets from a `kubeconfig`, the `Cluster` fields (Go and JSON) must
+be named the same as `"k8s.io/client-go/tools/clientcmd/api/v1".Cluster`, with the
+exception of `InsecureSkipTLSVerify`, which has been left out to discourage its use, and
+`CertificateAuthority`, which has been left out since CA data will always be passed
+to the plugin as bytes.
 
 This data is passed to the executable via the `KUBERNETES_EXEC_INFO` environment
 variable in a JSON serialized object.  Note that an environment variable is used
 over passing this information via standard input because standard input is
 reserved for interactive flows between the user and executable (i.e. to prompt
 for username and password).
 
+To make it easier for a plugin to use this `KUBERNETES_EXEC_INFO` environment
+variable to connect to the referent cluster, a set of helper functions will be added
+to create a `"k8s.io/client-go/rest".Config` from the `KUBERNETES_EXEC_INFO`
+environment variable. These helper functions will live in
+`k8s.io/client-go/tools/auth/exec/exec.go` so that non-plugin developers won't pull
+in this new package and the new package can safely depend on
+`k8s.io/client-go/rest`. The helper functions are as follows.
+
+```golang
+// LoadClusterFromEnv is a helper-wrapper around LoadCluster that loads from the well-known
+// KUBERNETES_EXEC_INFO environment variable. The KUBERNETES_EXEC_INFO environment
+// variable is expected to contains a serialized client.authentication.k8s.io ExecCredential
+// object.
+func LoadClusterFromEnv(into runtime.Object) (*rest.Config, error)
+
+// LoadCluster loads the configuration needed for an exec plugin to communicate with a
+// cluster. LoadCluster expects the provided data to be a serialized
+// client.authentication.k8s.io ExecCredential object.
+//
+// Specifically, LoadCluster loads
+//   1. a rest.Config from the ExecCredential.Spec.Cluster field and returns it, and
+//   2. a runtime.Object from the ExecCredential.Spec.Cluster.Config field into
+//      the provided into runtime.Object (into can be nil to ignore this field).
```

Comment:
would it be simpler to do
```
func LoadCluster(data []byte) (execCredential runtime.Object, *rest.Config, error)
```

and let them type-assert the version they expect and get their exec config from `execCredential.spec.cluster.config` themselves?

Reply:
Yes? Probably? You would have a better gut feel for this so I am tempted to go with the function signature that you proposed.

Your function signature gets rid of an error case where someone calls `LoadCluster(data, &corev1.Pod{})` on my function signature. So I think I do like your's better.

I suppose injecting references for return values is nice in situations where you want the caller to be able to allocate the memory necessary for an object, but I won't be able to outsmart the Go compiler here, and it's 2020 so why even worry about memory...